google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/read_rows/StateMachine.java

```diff
@@ -0,0 +1,414 @@
+import com.google.cloud.bigtable.data.v2.wrappers.RowAdapter.RowBuilder;
+import com.google.common.base.Preconditions;
+import com.google.protobuf.ByteString;
+import java.util.List;
+
+/**
+ * A state machine to merge {@link com.google.bigtable.v2.ReadRowsResponse.CellChunk}s from a
+ * ReadRows response stream. Building of the actual row object is delegated to a {@link RowBuilder}.
+ * This class is not thread safe.
+ *
+ * <p>The inputs are:
+ *
+ * <ul>
+ *   <li>CellChunks that must be merged to form logical rows
+ *   <li>Last scanned row notifications from the server to signal a more efficient resume point
+ *   <li>Row consumption events that reset the state machine for the next row
+ * </ul>
+ *
+ * <p>The outputs are:
+ *
+ * <ul>
+ *   <li>Logical rows that were constructed using the {@link RowBuilder}
+ *   <li>Special marker rows that represent resumption points that were sent by Last scanned row
+ *       notifications.
+ * </ul>
+ *
+ * <p>Expected Usage:
+ *
+ * <pre>{@code
+ * StateMachine stateMachine = new StateMachine<>(myRowAdapter);
+ * while(responseIterator.hasNext()) {
+ *   ReadRowsResponse response = responseIterator.next();
+ *   if (!response.getLastScannedRowKey().isEmpty()) {
+ *     stateMachine.handleLastScannedRow(response.getLastScannedRowKey());
+ *   }
+ *
+ *   for(CellChunk chunk : response.getChunksList()) {
+ *     boolean hasCompleteRow = stateMachine.handleChunk(chunk);
+ *     if (hasCompleteRow) {
+ *       MyRow row = stateMachine.consumeRow();
+ *
+ *       if (myRowAdapter.isScanMarkerRow(row) {
+ *         // do something with marker
+ *       } else {
+ *         // do something with row
+ *       }
+ *     }
+ *   }
+ * }
+ * }</pre>
+ *
+ * <p>Package-private for internal use.
+ *
+ * @param <RowT> The type of row the adapter will build
+ */
+final class StateMachine<RowT> {
+  private final RowBuilder<RowT> adapter;
+  private State currentState;
+  private ByteString lastCompleteRowKey;
+
+  // Track current cell attributes: protocol omits them when they are repeated
+  private ByteString rowKey;
+  private String familyName;
+  private ByteString qualifier;
+  private long timestamp;
+  private List<String> labels;
+  private long expectedCellSize;
+  private long remainingCellBytes;
+
+  private RowT completeRow;
+
+  /**
+   * Initialize a new state machine that's ready for a new row.
+   *
+   * @param adapter The adapter that will build the final row.
+   */
+  StateMachine(RowBuilder<RowT> adapter) {
+    this.adapter = adapter;
+    reset();
+  }
+
+  /**
+   * Feeds the last scanned serverside key into the state machine. The adapter will wrap this key in
+   * a special marker row that can be used downstream for efficient resume. It is the callers
+   * responsibility to eventually filter out this row.
+   *
+   * <dl>
+   *   <dt>Valid states:
+   *   <dd>{@link StateMachine#AWAITING_NEW_ROW}
+   *   <dt>Resulting states:
+   *   <dd>{@link StateMachine#AWAITING_ROW_CONSUME}
```

Comment:
I'm a bit confused about the purpose of this state and why it's what we transition to here...

Reply:
I updated the doc to clarify. But basically to prevent the caller accidently forgetting to consume produced data, the state machine goes into AWAITING_ROW_CONSUME state. The scan marker counts as a row to consume